devito/finite_differences/differentiable.py

```diff
@@ -0,0 +1,189 @@
+        merged["fd"] = dict(getattr(self, 'fd', {}), **getattr(other, 'fd', {}))
+        merged["staggered"] = self.staggered
+        merged["dtype"] = self.dtype
+        return merged
+
+    def __add__(self, other):
+        return Differentiable(sympy.Add(*[getattr(self, '_expr', self),
+                                          getattr(other, '_expr', other)]),
+                              **self._merge_fd_properties(other))
+
+    __iadd__ = __add__
+    __radd__ = __add__
+
+    def __sub__(self, other):
+        return Differentiable(sympy.Add(*[getattr(self, '_expr', self),
+                                          -getattr(other, '_expr', other)]),
+                              **self._merge_fd_properties(other))
+
+    def __rsub__(self, other):
+        return Differentiable(sympy.Add(*[-getattr(self, '_expr', self),
+                                          getattr(other, '_expr', other)]),
+                              **self._merge_fd_properties(other))
+
+    __isub__ = __sub__
+
+    def __mul__(self, other):
+        return Differentiable(sympy.Mul(*[getattr(self, '_expr', self),
+                                          getattr(other, '_expr', other)]),
+                              **self._merge_fd_properties(other))
+
+    __imul__ = __mul__
+    __rmul__ = __mul__
+
+    def __truediv__(self, other):
+        return Differentiable(sympy.Mul(*[getattr(self, '_expr', self),
+                                          getattr(other, '_expr', other)**(-1)]),
+                              **self._merge_fd_properties(other))
+
+    def __rtruediv__(self, other):
+        return Differentiable(sympy.Mul(*[getattr(other, '_expr', other),
+                                          getattr(self, '_expr', self)**(-1)]),
+                              **self._merge_fd_properties(other))
+
+    __floordiv__ = __truediv__
+    __rdiv__ = __rtruediv__
+    __div__ = __truediv__
+    __rfloordiv__ = __rtruediv__
+
+    def __pow__(self, exponent):
+        if exponent > 0:
+            return Differentiable(sympy.Mul(*[getattr(self, '_expr', self)]*exponent,
+                                            evaluate=False),
+                                  **self._kwargs)
+        elif exponent < 0:
+            return Differentiable(ExprDiv(sympy.Number(1),
+                                          sympy.Mul(*[getattr(self, '_expr',
+                                                      self)]*(-exponent),
+                                                    evaluate=False)),
+                                  **self._merge_fd_properties(None))
+        else:
+            return sympy.Number(1)
+
+    def __neg__(self):
+        return Differentiable(sympy.Mul(*[getattr(self, '_expr', self), -1]),
+                              **self._merge_fd_properties(None))
+
+    def __str__(self):
+        if self.is_Function:
+            return super(Differentiable, self).__str__()
+        return to_expr(self._expr).__str__()
```

Comment:
perhaps we don't need to wrap it with ``to_expr``

Reply:
probably, was getting some `Differentiable` in the str at some point so had to put it.